src/Build/Evaluation/ItemSpec.cs

```diff
@@ -504,4 +458,74 @@ public GlobFragment(string textFragment, string projectDirectory)
             && TextFragment[2] == '*'
             && FileUtilities.IsAnySlash(TextFragment[3]);
     }
+
+    internal sealed class MetadataSet<P, I> where P : class, IProperty where I : class, IItem, IMetadataTable
+    {
+        private readonly Dictionary<string, MetadataSet<P, I>> _children;
```

Comment:
Why use this convoluted data structure instead of dumping all metadata values from the Itemspec into a flat set of normalized metadata values?

Reply:
Still not 100% sure I agree with the data structure, but I want to clarify my understanding from the PR review today.

A flat set of normalized metadata values would prevent us from being able to determine exactly _which_ metadata values from _which_ items were matched.

We'd need to do this to enable the logic in a test like `RemoveWithMultipleItemReferenceOnMatchingMetadata`.
```xml
<I3 Remove='@(I1);@(I2)' MatchOnMetadata='M1' />
```
This remove attribute effectively translates to:
Remove items in I3 where:
- I1 **OR** I2 match on the following metadata:
   - M1

I believe this also enables something like:

```xml
<I3 Remove='@(I1);@(I2)' MatchOnMetadata='M1;M2' />
```
Where this remove attribute translates to:
Remove items in I3 where:
- I1 **OR** I2 match on the following metadata:
   - M1 **AND** M2

@Forgind is this understanding correct? If so I think there should be a test for multiple items and multiple metadata.

Reply:
Wrote detailed comment for MetadataTrie class. Will push soon.